samples/Kconfig

```diff
@@ -31,5 +31,67 @@ if LOG
 config LOG_DEFAULT_LEVEL
 	default 1 if NCS_SAMPLES_DEFAULTS
 
-
 endif # LOG
+
+# Openthread configuration features sets from Nordic Semiconductor
+
+choice
+	prompt "OpenThread features set"
+	depends on NET_L2_OPENTHREAD
+	help
+	  This option selects the Thread stack features set.
+
+config OPENTHREAD_FEATURES_MASTER
+	bool "Enable the complete set of OpenThread features"
+	imply OPENTHREAD_BORDER_AGENT
+	imply OPENTHREAD_BORDER_ROUTER
+	imply OPENTHREAD_CHILD_SUPERVISION
+	imply OPENTHREAD_COAP
+	imply OPENTHREAD_COAPS
+	imply OPENTHREAD_COMMISSIONER
+	imply OPENTHREAD_DIAG
+	imply OPENTHREAD_DNS_CLIENT
+	imply OPENTHREAD_DHCP6_SERVER
+	imply OPENTHREAD_DHCP6_CLIENT
+	imply OPENTHREAD_ECDSA
+	imply OPENTHREAD_ENABLE_SERVICE
+	imply OPENTHREAD_IP6_FRAGM
+	imply OPENTHREAD_JAM_DETECTION
+	imply OPENTHREAD_JOINER
+	imply OPENTHREAD_MAC_FILTER
+	imply OPENTHREAD_MTD_NETDIAG
+	imply OPENTHREAD_RAW
+	imply OPENTHREAD_SLAAC
+	imply OPENTHREAD_SNTP_CLIENT
+	imply OPENTHREAD_UDP_FORWARD
+config OPENTHREAD_FEATURES_OPTIMIZED_FTD
+	bool "Enable optimized OpenThread features for FTD"
+	imply OPENTHREAD_CHILD_SUPERVISION
+	imply OPENTHREAD_COAP
+	imply OPENTHREAD_COAPS
+	imply OPENTHREAD_DNS_CLIENT
+	imply OPENTHREAD_DHCP6_CLIENT
+	imply OPENTHREAD_ECDSA
+	imply OPENTHREAD_ENABLE_SERVICE
+	imply OPENTHREAD_IP6_FRAGM
+	imply OPENTHREAD_JAM_DETECTION
+	imply OPENTHREAD_JOINER
+	imply OPENTHREAD_MAC_FILTER
+	imply OPENTHREAD_SLAAC
+	imply OPENTHREAD_SNTP_CLIENT
+	imply OPENTHREAD_UDP_FORWARD
+config OPENTHREAD_FEATURES_OPTIMIZED_MTD
+	bool "Enable optimized OpenThread features for MTD"
+	imply OPENTHREAD_CHILD_SUPERVISION
+	imply OPENTHREAD_COAP
+	imply OPENTHREAD_COAPS
+	imply OPENTHREAD_DNS_CLIENT
+	imply OPENTHREAD_DHCP6_CLIENT
+	imply OPENTHREAD_ECDSA
+	imply OPENTHREAD_IP6_FRAGM
+	imply OPENTHREAD_JAM_DETECTION
+	imply OPENTHREAD_JOINER
+	imply OPENTHREAD_MAC_FILTER
```

Comment:
Why is all of this in `samples/Kconfig` ?
This seems like a very bad choice, as everything here is openthread specific.

This configuration itself looks like something that ought to be in an `overlay-openthread-ftd.conf`, `overlay- openthread-mtd.conf` and so on.

If we should start placing such groups everywhere in Kconfig menus, it will clutter it a lot.
I would therefore like a wider discussion if this is a principle we want to start using.

Has overlay files been considered instead ?
Overlay files are a commonly used pattern.

@carlescufi @mbolivar-nordic @rlubos  thoughts ?

Reply:
also remember, the benefit of an overlay file is that users can easily copy that file into their own application/repo and adjust settings (remove a single setting or adjust a value), and thus have openthread settings grouped together in a conf file.
And by providing an overlay file, then users will learn about this possibility.

That is  harder when such settings are set through `imply` in Kconfig as well as users may never be aware of the overlay possibility.

Reply:
The overlay files approach was discarded since there was some opposition within the Thread Team to the user having to type the whole `-DCONF_FILE="prj.conf overlay-openthread-ftd.conf"` every time they want to build a sample.

That said I understand your concerns. Maybe we could find a way to provide common configuration for similar samples? Like the `samples/openthread/common/` proposed by @rlubos.

Reply:
> The overlay files approach was discarded since there was some opposition within the Thread Team to the user having to type the whole -DCONF_FILE="prj.conf overlay-openthread-ftd.conf" every time they want to build a sample.

So instead they would need to do `ninja menuconfig` and then select `OPENTHREAD_FEATURES_OPTIMIZED_MTD=y` or how is the proposed workflow ?

I fail to see why entering `menuconfig` is easier than writing: `-DOVERLAY_CONFIG=overlay-openthread-ftd.conf`.
Note, using `-DOVERLAY_CONFIG=<file>` then you don't need to specify `prf.conf`.

Remember `-DOVERLAY_CONFIG` is a common principle, so I can easily find all overlay files (bt, mbedtls, openthread, ...) by a simple search.
But openthread (this PR) deviates from this, and I need a good reason for that.

Reply:
Well, the idea of this PR is being able to select the features set in the `prj.conf` file for each sample and having a common single place for the OpenThread configuration. I'm not opposed to your overlays proposal, but we should find a good place to store them.

Reply:
>  I'm not opposed to your overlays proposal, but we should find a good place to store them.

agreed, and i'm trying to understand your needs, or where the overlay feature could be improved :wink:


The `prj.conf` can only select a single of those the groups at a time, so if wanting to use `MTD` instead of `FTD`, then that would be a change in the `prj.conf`, is that correctly understood.

Reply:
Yes, you understood correctly.

Reply:
And maybe it's not mentioned here, but eventually there is a goal to be able to pick one of these feature sets in order to achieve Thread certification by inheritance, as it was possible with nRF5 SDK:
- https://infocenter.nordicsemi.com/topic/sdk_tz_v4.1.0/thread_certification.html
- https://infocenter.nordicsemi.com/topic/sdk_tz_v4.1.0/thread_openthread_libraries.html

Reply:
> Yes, you understood correctly.

Good. And thanks for the links regarding certification.

But I do think obtain Thread Certification by inheritance can be ensured, both with your proposed solution but also with a `-DOVERLAY_CONFIG=overlay-openthread-ftd.conf` approach.

But it's good to have the direction in mind.

I have considered this proposal some more, and the are a couple of major reasons why I still dislike this approach:
1) It deviates from existing principle of overlay config.
1) This principle outlined in the PR makes it much harder for end-user to see which settings are related to `OPENTHREAD_FEATURES_OPTIMIZED_FTD`. 
    With an overlay file, a user can easily examine it, like:
    ```
    $ cat overlay-openthread-ftd.conf
    OPENTHREAD_CHILD_SUPERVISION=y
    OPENTHREAD_COAP=y
    OPENTHREAD_COAPS=y
    OPENTHREAD_DNS_CLIENT=y
    OPENTHREAD_DHCP6_CLIENT=y
    ....
    ```
    but the Kconfig file content can be located in any Kconfig file. In this case, `samples/Kconfig`.
    But that file contains much more than just OpenThread.
    The `.config` is of little help, cause it cannot be determined which settings in `.config` that are a result of `OPENTHREAD_FEATURES_OPTIMIZED_FTD`.
    User can of course use `menuconfig` and search for `OPENTHREAD_FEATURES_OPTIMIZED_FTD` and then use help description to see more info. But that is not user friendly, and definitely not for a newcomer to NCS.
1) If the principle in this PR is expended to other areas, then we will start to clutter Kconfigs with such settings everywhere.
    This will also make it even harder for everyone. Like, OpenThread --> Look in Kconfig files. Network --> Look in overlay files.

Another concern is the use of `imply` because it doesn't a guarantee that the value will be set.
And if the setting is not guaranteed, then I would say it makes `Thread Certification by inheritance` problematic. 

But I do see there might be a need for easier inclusion of an overlay file into `prj.conf`.
For boards, this is already in place, so if you have:
```
my_sample
|--- prj.conf
|--- boards
        |--- nrf52840dk_nrf52840.conf
        |--- nrf52840dk_nrf52811.conf
```
then building for board `nrf52840dk_nrf52840` will automatically merge `prj.conf` and `nrf52840dk_nrf52840.conf`, but if building `nrf9160dk_nrf9160`, then only `prj.conf` is used.

So instead of the Kconfig proposal, I would suggest we look into a possibility of making it easier to include an overlay file into `prj.conf`.
Just for the example: `prj.conf`
```
#include "overlay-openthread-ftd.conf"
CONFIG_SOMETHING_ELSE=y
...
```

Reply:
This discussion (and @ru-fu -- thanks!) reminded me about Build types we have already documented: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/gs_modifying.html#gs-modifying-build-types

Not my field of specialization, but shouldn't we be using this feature? If not, we need to clearly explain in the docs what the differences with build types are (and what the advantages of the configuration sets in the present form are).

Reply:
@tejlmand being able to include overaly files in the `prj.conf` files would be great. However, I don't see the way (https://docs.zephyrproject.org/latest/guides/kconfig/setting.html) and coming up with a new system would be out of scope of this PR.

On the contrary, we already have precedent of using Kconfig for this purpose with `CONFIG_NCS_SAMPLES_DEFAULTS`, which is widely used in the samples.

That's the approach of my latest commit. I know it deviates from the overlays approach but I didn't find any other better solution. Maybe providing a good documentation for this would help users not to get confused. Also those Kconfig sections could be moved to the `samples/openthread` folder.

Reply:
Well, to be honest I'm not sure if people have considered the wider impact of that PR: https://github.com/nrfconnect/sdk-nrf/pull/1594

There is also a slight difference in the fact that that is intended to be default for all samples.
But as this seems to be a default for all NCS samples, then I don't understand the reason for this code:
https://github.com/rakons/fw-nrfconnect-nrf/blob/d4c828e7ad4ca6252140ec1d86e13f6c8debca78/samples/Kconfig#L7-L15
```
config NCS_SAMPLES_DEFAULTS
	bool "Enable NCS sample default settings"
	default n
	imply LOG
	imply LOG_MINIMAL
	imply ASSERT
	imply HW_STACK_PROTECTION
	help
	  Use the default configuration for NCS samples.
```
cause there is already a `Kconfig.defconfig` principle in place, which is used by SoCs.
Like here:
https://github.com/zephyrproject-rtos/zephyr/blob/master/soc/arm/nordic_nrf/Kconfig.defconfig#L6-L40

So why none of the reviewers in #1594 didn't suggest to put that code into a `Kconfig.samples.defconfig` with following content instead:
```
config LOG
    default y

config LOG_MINIMAL
    default y

config ASSERT
    default y
```
is unclear to me. 
Note, above code would most likely have to be surrounded by `if <SOME_FLAG>` to only be effective for NCS samples.

Then you could also avoid:
```
CONFIG_NCS_SAMPLES_DEFAULTS=y
```
in all files.

But it seems I was not part of that review.

Reply:
> @tejlmand being able to include overaly files in the prj.conf files would be great. However, I don't see the way (https://docs.zephyrproject.org/latest/guides/kconfig/setting.html) and coming up with a new system would be out of scope of this PR.

Maybe out of scope for this PR, but if extensions are needed to give a better user experience, and ensure consistency on where things are located then we should always work on such solutions.

Reply:
@tejlmand @rlubos @carlescufi 
Could the `source` or `rsource` [Kconfig extensions](https://docs.zephyrproject.org/latest/guides/kconfig/extensions.html) be used with some kind of symbol expansion? Something like:

`source "$(CONFIG_INCLUDE_CONF)"`

Being the `CONFIG_INCLUDE_CONF' an available configuration string.

Actually I tried this but the expansion doesn't work in that way, but maybe someone is aware of a similar way. Also it would be nice to be able to invoke this `source` more than once for different files.

Reply:
that will only work inside Kconfig, and not in the conf files.
I'm not sure such a suggestion would be accepted upstream, as that would change the format of the file, compared to Linux kernel.
I have made a small proposal here that you can test:
https://github.com/nrfconnect/sdk-zephyr/pull/347
```
cd $ZEPHYR_BASE
git fetch https://github.com/nrfconnect/sdk-zephyr pull/347/head
git cherry-pick 727c7a82f7 05344053e2
```

and then you can update your `prf.conf` with:
```
#include "${ZEPHYR_NRF_MODULES_DIR}/samples/openthread/ot-overlay.conf"
....
```

Try it out, and give feedback.

Reply:
Thanks a lot @tejlmand , it works as a charm!

It should be `ZEPHYR_NRF_MODULE_DIR` thought.

Reply:
> Thanks a lot @tejlmand , it works as a charm!
> 
> It should be `ZEPHYR_NRF_MODULE_DIR` thought.

Right, fixed ^^^